### ISSUE TITLE
Gnocchi does not cleanup deleted resources

### ISSUE THREAD

**Issue by sombrafam** [gnocchixyz/gnocchi#1050]
The resources that gnocchi creates are always related to resources in some service in OpenStack (let's say a instance). When the instance (or other resource) is deleted in the service, gnocchi keeps all its data in both mysql and storage backend.

This might be a problem when dealing with systems that do a lot of creation/deletion of resources.

Version: 4.2.5

How to reproduce your problem:
    - Configure ceilometer+gnocchi
    - Create resources on the system (volumes, instances, etc)
    - Delete the resources
    - Check the gnocchi resources and metrics list

What is the result that you get: Resources deleted in OpenStack still are listed in gnocchi
What is result that you expected: Resources deleted in OpenStack still are *not* listed in gnocchi

**Comment by sombrafam** [comment]
From what I understood, gnnochi does not mark then as deleted, at least not in the configuration I'm using. What I needed to do was to cross the resource ids with nova/cinder/glance and then set the 'metric.delete' field. Then, gnocchi deleted the metadata and data.

```SQL
# Total metric count
select count(id) from gnocchi.metric;

# Total resource count
select count(id) from gnocchi.resource;

use gnocchi;
################################################################################
# Instance related cleanups
################################################################################
update gnocchi.metric as upm
set upm.status='delete'
where upm.id in
(select m.id
from nova.instances i, gnocchi.resource r, (select * from gnocchi.metric) m
where i.deleted <> 0
  and r.original_resource_id like concat('%', i.uuid,'%')
  and m.resource_id = r.id);

# Delete dangling instance resources
delete from resource where id in (select r.id
from nova.instances i, (select * from gnocchi.resource) r
where i.deleted <> 0
  and r.original_resource_id like concat('%', i.uuid,'%'));

################################################################################
# Volumes cleanups
################################################################################
# Delete metrics that should be deleted
update gnocchi.metric
set gnocchi.metric.status='delete'
where gnocchi.metric.id in (
select id from (select id,resource_id from metric) as pruned_metric where resource_id in (
    select id from resource where type='volume' and original_resource_id in (
        select id from cinder.volumes where deleted <> 0 union select id from cinder.snapshots where deleted <> 0)));

# Delete dangling volume resources
delete from resource where original_resource_id in (
  select id from cinder.volumes where deleted <> 0
  union select id from cinder.snapshots where deleted <> 0);

################################################################################
# Images cleanups
################################################################################
# Delete metrics that should be deleted
update gnocchi.metric
set gnocchi.metric.status='delete'
where gnocchi.metric.id in (
select id from (select id,resource_id from metric) as pruned_metric where resource_id in (
    select id from resource where type='image' and original_resource_id in (
        select id from glance.images where deleted <> 0)));

# Delete dangling image resources
delete from resource where original_resource_id in (select id from glance.images where deleted <> 0);

# Total metric count
select count(id) from gnocchi.metric;

# Total resource count
select count(id) from gnocchi.resource;
```

**Comment by sombrafam** [comment]
FYI, just to confirm. I checked the code and there's no garbage collector.

**Comment by chungg** [comment]
you probably shouldn't have deleted the records from the database. the janitor process[1] uses that reference to cleanup data. so you effectively deleted the metadata but all the numerical data associated with it will remain on your storage indefinitely.

[1] https://github.com/gnocchixyz/gnocchi/blob/7a3ddc4ceb64bf648a5f383814632f72b8700a8f/gnocchi/cli/metricd.py#L262-L271

**Comment by sombrafam** [comment]
I'm not actually deleting any metric see that "set gnocchi.metric.status='delete'". I'm just setting then as delete so the janitor will clean up. The things I delete are the resources which are not being collected.

**Comment by jd** [comment]
I think the problem is clear from the first sentence:

> The resources that gnocchi creates are always related to resources in some service in OpenStack (let's say a instance). When the instance (or other resource) is deleted in the service, gnocchi keeps all its data in both mysql and storage backend.

IIUC @sombrafam expects resources to be deleted from Gnocchi when they are deleted from e.g. OpenStack Nova.

This is not the case, it never has been, and it'll never be.

**Comment by sombrafam** [comment]
@jd what is the point of closing this bug then? I know, the project is not maintained, but it does necessarily not means it will stay like that forever ;)

**Comment by jd** [comment]
This is not a feature we ever wanted to implement in Gnocchi. This is a timeseries database, it's up to the API users to manage the resource lifecycle.

There's already an API to delete resources from Gnocchi.

You don't expect e.g. MySQL to delete your customer records when they are deleted from your Salesforce service (or vice-versa).

If you want OpenStack to reflect the resource existence/deletion in Gnocchi, then you should ask OpenStack (probably Ceilometer) to implement such a feature.

**Comment by sombrafam** [comment]
I think I see your point. One question, we are not seeing object being deleted after their expiration, I didn't find that part in the code also. Do you know if that is implemented?

**Comment by jd** [comment]
"objects"?

If you mean resources: https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/rest/api.py#L1045-L1055

There's also a batch deletion for resources: https://github.com/gnocchixyz/gnocchi/blob/master/gnocchi/rest/api.py#L1154-L1183

For metrics, you can look at the same file + the janitor code that @chungg already shared — deletion is async in order to delete metrics from the object storage.

**Comment by sombrafam** [comment]
Yes, that part I got. It's how you delete an object using the API. My question is, the archive policy has an timespan, which I believe shold be the TTL of the object right? After that is expired, the object should be deleted right (where)? We have seem several objects much older then their timespan.

**Comment by sombrafam** [comment]
hmmm, well that is super misleading, I can tell as when I firstly started to learn about gnocchi the TTL meaning was the obvious thing it came to me. And a lot of operators I have talked have been lead to think that. 
How do you guys feel about having a 'ttl' field in the policy and implement a collector in the Janitor?

**Comment by jd** [comment]
Feel free to open an issue for a feature request.

**Comment by ahmedramzy14** [comment]
Dears,

Last 2 days i faced health_warn on my ceph storage cluster(pool metrics has many more objects per pg than average (to few pgs?))

That's lead to most of my VMs Flactuated(can't connected to their disks for a while), because metrics pool has I/O slow requests affected Volume pool 

Ceph health be OK after i have deleted Gnocchi resources (was more than 3 thousand Gnocchi resources)

Also, metrics objects be 194297 (was more than 6 million)

And advise? 
And how i can know the RCA to avoid this issue in future.

Thanks

**Comment by sombrafam** [comment]
Hey,

As you may have seen in this discussion, Gnocchi does not provide a cleanup
routine for old objects.
You need to clean them either via API or in a dirty way through SQL.

This[1] cleans dangling (resources that was deleted from OpenStack) objects
from Gnocchi. If you want give it a try and add to cron.

[1] https://gist.github.com/sombrafam/a3259f2f8855194bdb4ffae743ddc0be

Em sáb, 26 de out de 2019 às 08:17, ahmedramzy14 <notifications@github.com>
escreveu:

**Comment by engel75** [comment]
So to put it more simply, the need for storage continues to grow?
no cleanup at all?

**Comment by DEmmons** [comment]
While this issue is rightly closed, I am adding a brief update as it continues to be relevant in 2026, since Ceilometer does not implement a periodic cleanup utility, and many deployments no longer set Gnocchi as an event publisher. This has the result that the intended batch delete method in the Gnocchi REST API cannot be used, because the ended_at property does not get set on resources that are deleted from OpenStack. This can be fixed simply by adding "- gnocchi://" in the publishers section of the event_pipeline.yaml file, on the machines running the Ceilometer notification agent. 

With this property being set by Ceilometer, the batch delete method can be used to clean up all stale measures stored in Ceph, as well as the resource and metric entries in Gnocchi's database, with one command. This method also allows keeping measures for deleted resources that have not reached the end of their retention yet. As an example, default archive policies usually set a retention of 30 days, so the following API call (utilizing the OpenStack CLI client) would be used to clean up all resources deleted more than 30 days ago:

`openstack metric resource batch delete '(ended_at < "-30day")'`

This can be implemented as a cron job as well. Note that the more complicated SQL commands described previously would still be needed once, to clean up resources deleted before the event publisher was set in Ceilometer, if any are present.

**Comment by DEmmons** [comment]
Another update on this:
It looks like a solution to the need for a cron job or similar cleanup job was also committed last year, and is available in 4.7.0 as described in this PR and the ones related to it:
https://github.com/gnocchixyz/gnocchi/pull/1450

This enables setting an expiration age in Gnocchi configs for deleted resources to automatically be cleaned after reaching. Even better, it includes detection for measures that have not recieved new data in that period and cleans those up too. This looks like it should address virtually all the issues discussed in this issue, though users with multiple archive policies containing differing retentions may still want to set up a more fine-tuned cleanup (otherwise it would neet to be set to the value of the archive policy with the longest retention and that would be used for all of them)
